src/number_theory/legendre_symbol/auxiliary.lean

```diff
@@ -30,18 +34,194 @@ begin
     { exact nat.odd_of_mod_four_eq_three h }, },
 end
 
+-- Auxiliary stuff for monoids
+
+/-- If `x ^ n = 1`, then `x ^ m` is the same as `x ^ (m % n)` -/
+-- SUGGESTION: move this to `algebra.group_power.basic`, right after `pow_sub_mul_pow`
+@[to_additive nsmul_eq_mod_nsmul "If `n • x = 0`, then `m • x` is the same as `(m % n) • x`"]
+lemma pow_eq_pow_mod {M : Type*} [monoid M] {x : M} (m : ℕ) {n : ℕ} (h : x ^ n = 1) :
+  x ^ m = x ^ (m % n) :=
+begin
+  nth_rewrite 0 [← nat.div_add_mod m n],
+  rw [pow_add, pow_mul, h, one_pow, one_mul],
+end
+
+/-- We define the inverse as a `monoid_with_zero_hom` by extending the inverse map by zero
+on non-units. -/
+-- This exists for groups with zero (`inv_monoid_with_zero_hom`)
+-- SUGGESTION: put the following two definitions in `algebra.group_with_zero.basic`,
+-- before `inv_monoid_with_zero_hom`
+noncomputable
+def comm_monoid_with_zero.inv_hom {M : Type*} [comm_monoid_with_zero M] :
+  M →*₀ M :=
```

Comment:
This is NOT the same as `inv_monoid_with_zero_hom` because it uses `ring.inverse`, not `inv`. Please call it `monoid_with_zero_hom.inverse`, `or `inverse_monoid_with_zero_hom` and put it where you said.

Reply:
OK; I prefer `monoid_with_zero_hom.inverse`.